### PR TITLE
Add functionality to Spritesheet

### PR DESCRIPTION
**Pull request by JonnyPtn**
Adds some functionality required by my editor:

`SaveToFile(const std::string& path)` - Saves the sprite sheet at the given path
`getSprites()` - Returns the map of sprites in this spritesheet
`getTexturePath()` - Returns the path to the texture used by this spritesheet

I also added `m_texturePath` and `m_smooth` member variables for persistence.

It's likely I'll be adding more to this in future, so it's up to you if you want to leave this open until it's more "complete" or just merge the incremental changes as they come through

edit: I just remembered, This also makes the "src" parameter relative to the location of the sprite file. previous behaviour was just relative to the application's working directory, which could be anywhere

### CODE REVIEW

File: xyginext/src/graphics/SpriteSheet.cpp [hunk 1]
```diff
@@ -90,7 +91,8 @@ bool SpriteSheet::loadFromFile(const std::string& path, TextureResource& texture
             }
 
             Sprite spriteComponent;
-            spriteComponent.setTexture(*texture);
+            if (texture != NULL)
```
Comment (fallahn) [inline xyginext/src/graphics/SpriteSheet.cpp]: nullptr please :) I personally prefer braces around the statement too,  just for consistency

Comment (JonnyPtn) [reply xyginext/src/graphics/SpriteSheet.cpp]: No problem, will be changed in next commit

**Comment by JonnyPtn** [comment]
I've made some more changes to this, but have hit a bit of a block with how animations are handled. As I understand it, the animation name is only ever available to the spritesheet class, and after that has been used to load the sprites/animations, they can only be referenced by index, which I'm required to manually match up with the order of animations in the sprite sheet, which leaves lots of room for error.

Locally, I've added a fixed size char array to xy::Sprite::Animation which is set to the name specified in the sprite sheet file, which is a fairly simple/quick solution, but I also pondered with the idea of separating Animation from Sprite, and making Animation a more abstract concept, which could be used for animating things other than bounds (e.g. transform, colour, etc.), thoughts...?

**Comment by JonnyPtn** [comment]
I've added an ID to xy::Sprite::Animation now, along with a few other sprite sheet helpers

**Comment by fallahn** [comment]
As always the hard work is very much appreciated. Should I assign you to #11 ? 😊

**Comment by JonnyPtn** [comment]
I've updated this now to include `setSprite` to assign a sprite in the sprite sheet

**Comment by JonnyPtn** [comment]
I've added more functionality to SpriteAnimation to get/set the current frame of the animation. My requirement was for the editor to allow frame-by-frame editing, but I'm offering it upstream as I think It would be useful in xygine games too (I'm trying to avoid including things purely because the editor wants them).

If you want the sprite/animation changes separated into a separate PR let me know

**Comment by fallahn** [comment]
Aha, this is actually something I had need for at one point but didn't have the time to implement myself. Awesome!

**Comment by JonnyPtn** [comment]
I've started to think maybe making the texture path relative to the sprite file might not be the best Idea, but I'm not really sure of the best alternative...

I think what I might do is add the concept of a "project" to the editor, rather than editing the sprite files individually, so the project would be running from the same working folder as the game would be. It was something I was planning on doing in the future anyway, just not yet. what do you think?

**Comment by fallahn** [comment]
In previous attempts I did something along the lines of that - there was a project file which contained configuration details such as the working directory. It also meant it was quite easy to configure game specific settings, much like an [fgd](https://developer.valvesoftware.com/wiki/FGD) file defines entities for valve's hammer editor, although far less complex. The ConfigFile format should probably suffice

**Comment by JonnyPtn** [comment]
I've changed that back now, so texture paths aren't relative to sprite file. Fairly convinced this is the correct decision for the Time being, as I can't think of a reason for them to be relative except for the editor

**Comment by JonnyPtn** [comment]
ci is failing here as the travis machine can't find gcc-7 :s

seen this a few times before and not sure what's happening, seems to be an identical bistro to previous successful builds...

**Comment by fallahn** [comment]
Yeah travis seems to be having a few problems lately. I thought I'd just give it some time to see if they sort themselves out. (Seems to have passed, as of posting)

**Comment by JonnyPtn** [comment]
I'm beginning to think some of the further changes I had planned for this (additional behaviour for sprites and animations) should be done in a separate issue/PR so they can be separately discussed and tested.

To summarise what's included in this PR:

- Sprite::Animation has an id now
- SpriteAnimation can get/set current frame
- Spritesheets can be modified and saved to files

So If you're happy with this PR as it is, then I'd say it's ready to merge

**Pull request merged**